Fix stopping criteria for generation with inputs_embeds with dynamic shapes

skavulya opened the pull request:
# What does this PR do?
Fixes index out-of-bounds error which occurs when generating text from inputs_embeds.

```
import torch
from transformers import AutoModelForCausalLM, AutoTokenizer
from optimum.habana.transformers.modeling_utils import adapt_transformers_to_gaudi

adapt_transformers_to_gaudi()
model = AutoModelForCausalLM.from_pretrained(
    "gpt2",
).to('hpu')
tokenizer = AutoTokenizer.from_pretrained("gpt2")
model.generation_config.static_shapes = False
text = "Deepmind is a company"
input_ids = tokenizer.encode(text, return_tensors="pt").to("hpu")
inputs_embeds = model.get_input_embeddings()(input_ids)
outputs = model.generate(inputs_embeds=inputs_embeds, max_new_tokens=10)
print(
    "\noutputs:",
    tokenizer.decode(outputs[0], skip_special_tokens=True),
)
```

```
Traceback (most recent call last):
  File "/root/optimum-habana/test_embeds.py", line 20, in <module>
    outputs = model.generate(inputs_embeds=inputs_embeds, max_new_tokens=10)
  File "/usr/local/lib/python3.10/dist-packages/torch/utils/_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
  File "/root/optimum-habana/optimum/habana/transformers/generation/utils.py", line 1122, in generate
    result = self._greedy_search(
  File "/root/optimum-habana/optimum/habana/transformers/generation/utils.py", line 1862, in _greedy_search
    unfinished_sequences = unfinished_sequences & ~stopping_criteria(
  File "/root/optimum-habana/optimum/habana/transformers/generation/stopping_criteria.py", line 113, in gaudi_StoppingCriteriaList_call
    is_done = is_done | criteria(input_ids, scores, **kwargs)
  File "/root/optimum-habana/optimum/habana/transformers/generation/stopping_criteria.py", line 84, in gaudi_EosTokenCriteria_call
    is_done = torch.isin(input_ids[:, token_idx - 1], self.eos_token_id)
IndexError: index 5 is out of bounds for dimension 1 with size 1
```
<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

<!-- Remove if not applicable -->

Fixes # (issue)


## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

ssarkar2 reviewed (commented):
`python run_generation.py --model_name_or_path openai-community/gpt2 --use_kv_cache --max_new_tokens 128 --max_input_tokens=128 --batch_size 4 --use_hpu_graph --bf16 --no-ignore_eos` 

running on this branch, results of couple of runs:
2018, 2131, 2078, 2064
avg: 2072.7

on main couple of runs:
2225.31, 2255.12, 2346, 2261
avg: 2271.8


curlen and token_idx are the same number (for my test mentioned in this comment), except that token_idx is a tensor and curlen is a simple integer.
I suspect this difference between them causes the performance difference. If I replace token_idx to curlen in this branch i get back original perf as seen on main.. though of course then your test fails.

We need to make sure perf doesnt degrades, as well as the corner case you have mentioned passes.

ssarkar2 commented on optimum/habana/transformers/generation/utils.py:
```diff
@@ -1856,11 +1859,11 @@ def _greedy_search(
             if ignore_eos:
                 this_peer_finished = stopping_criteria(
-                    input_ids, scores, token_idx=cur_len, ignore_eos=ignore_eos, eos_token_id=eos_token_id
+                    input_ids, scores, token_idx=token_idx, ignore_eos=ignore_eos, eos_token_id=eos_token_id
```
stop_tkn_idx = token_idx if token_idx is None else cur_len

stopping_criteria(input_ids, scores, token_idx=stop_tkn_idx , ignore_eos=ignore_eos, eos_token_id=eos_token_id)

skavulya commented:
@ssarkar2 I have updated the PR to fix the performance regression by using a scalar `stop_tkn_idx`. Please let me know if you observe similar performance on your end.

The performance of `python run_generation.py --model_name_or_path openai-community/gpt2 --use_kv_cache --max_new_tokens 128 --max_input_tokens=128 --batch_size 4 --use_hpu_graph --bf16 --no-ignore_eos`

This PR:
 2165.66, 2248.42, 2202.1112 tokens/sec

Main
 2175.64, 2203.87, 2225.53 tokens/sec

I have also added  `get_final_stopping_criteria` to the `ignore_eos` case to prevent failures when `batch_size > 1` and `token_idx is None`. The stopping criteria returns a tensor whose shape is batch size. `get_final_stopping_criteria` converts it to a boolean.
https://github.com/huggingface/optimum-habana/blob/977f40eaaa91ab3f0ff804b754219f1f0fafbb86/optimum/habana/transformers/generation/utils.py#L1865-L1866

skavulya commented:
@ssarkar2 I merged the changes from this branch to https://github.com/huggingface/optimum-habana/pull/948 since they are related. Please review the changes there. Thanks.
